mlflow/utils/file_utils.py

```diff
@@ -588,6 +596,102 @@ def download_file_using_http_uri(http_uri, download_path, chunk_size=100000000,
+    # before sending to the workers so they can each individually seek to their respective positions
+    # and write chunks without overwriting.
+    open(download_path, "w").close()
+    starting_index = 0
+    if uri_type == ArtifactCredentialType.GCP_SIGNED_URL or uri_type is None:
+        # GCP files could be transcoded, in which case the range header is ignored.
+        # Test if this is the case by downloading one chunk and seeing if it's larger than the
+        # requested size. If yes, let that be the file; if not, continue downloading more chunks.
+        download_chunk(
+            range_start=0,
+            range_end=chunk_size - 1,
+            headers=headers,
+            download_path=download_path,
+            http_uri=http_uri,
+        )
+        downloaded_size = os.path.getsize(download_path)
+        # If downloaded size was equal to the chunk size it would have been downloaded serially,
+        # so we don't need to consider this here
+        if downloaded_size > chunk_size:
+            return {}
+        else:
+            starting_index = 1
+
+    failed_downloads = {}
+
+    def run_download(range_start, range_end):
+        range_start = i * chunk_size
+        range_end = range_start + chunk_size - 1
+
+        with tempfile.TemporaryFile(mode="w+") as temp_file:
+            download_proc = _exec_cmd(
+                cmd=[
+                    sys.executable,
+                    download_cloud_file_chunk.__file__,
+                    "--range-start",
+                    range_start,
+                    "--range-end",
+                    range_end,
+                    "--headers",
+                    json.dumps(headers or {}),
+                    "--download-path",
+                    download_path,
+                    "--http-uri",
+                    http_uri,
+                    "--temp-file",
+                    temp_file,
+                ],
+                throw_on_error=True,
+                synchronous=False,
+                capture_output=True,
+                stream_output=False,
+                env=env,
+            )
+            return_code = download_proc.wait()
+            if return_code != 0:
+                temp_file.seek(0)
+                file_contents = temp_file.read()
+                if file_contents:
+                    failed_downloads[i] = json.loads(temp_file.read())
+                else:
+                    raise Exception(
+                        "Error from download_cloud_file_chunk not captured, "
+                        f"return code {return_code}"
+                    )
+
+    with ThreadPoolExecutor(max_workers=MAX_PARALLEL_DOWNLOAD_WORKERS) as p:
+        for i in range(starting_index, num_requests):
+            range_start = i * chunk_size
+            range_end = range_start + chunk_size - 1
+            future = p.submit(run_download, range_start, range_end)
+            try:
+                future.result()
```

Comment:
What the current code does is:

- Upload a part
- Wait
- Upload a part
- Wait
...

It's not parallelized at all. The example in the official doc might help:
https://docs.python.org/3/library/concurrent.futures.html#threadpoolexecutor-example

Reply:
Thanks!